lib/jazzy/sourcekitten.rb

```diff
@@ -388,32 +400,63 @@ def self.filter_excluded_files(json)
       end.compact
     end
 
-    def self.names_and_urls(docs)
-      docs.flat_map do |doc|
-        # FIXME: autolink more than just top-level items.
-        [{ name: doc.name, url: doc.url }] # + names_and_urls(doc.children)
+    def self.name_match(name_part, docs)
+      return nil unless name_part
+      wildcard_expansion = Regexp.escape(name_part).gsub('\.\.\.', '[^)]*')
+      whole_name_pat = /\A#{wildcard_expansion}\Z/
+      docs.find do |doc|
+        whole_name_pat =~ doc.name
+      end
+    end
+
+    # Find the first ancestor of doc whose name matches name_part.
+    def self.ancestor_name_match(name_part, doc)
+      doc.namespace_ancestors.reverse_each do |ancestor|
+        if match = name_match(name_part, ancestor.children)
+          return match
+        end
+      end
+      nil
+    end
+
+    def self.name_traversal(name_parts, doc)
+      while doc && !name_parts.empty?
+        next_part = name_parts.shift
+        doc = name_match(next_part, doc.children)
       end
+      doc
     end
 
-    def self.autolink_text(text, data, url)
-      regex = /\b(#{data.map { |d| Regexp.escape(d[:name]) }.join('|')})\b/
-      text.gsub(regex) do
-        name = Regexp.last_match(1)
-        auto_url = data.find { |d| d[:name] == name }[:url]
-        if auto_url == url # docs shouldn't autolink to themselves
-          Regexp.last_match(0)
+    def self.autolink_text(text, doc, root_decls)
+      text.gsub(%r{<code>[ \t]*([^\s]+)[ \t]*</code>}) do
+        original = Regexp.last_match(0)
+        raw_name = Regexp.last_match(1)
+        parts = raw_name
```

Comment:
It'd be nice to match Objective-C square bracket syntax too: `[RLMRealm defaultRealm]`, `+[RLMRealm defaultRealm]` and `-[RLMRealm writeCopyToPath:encryptionKey:error:]`.